graphql-upload: Update types to v15

rdsfj opened the pull request:
Hello everyone 🙋‍♀️ 

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/jaydenseric/graphql-upload/blob/master/changelog.md#1500
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.


<br>
I've updated the types of `graphql-upload` to its version 15 as linked above. v15 is not the latest version, but still in use so it's necessary to haves the types available.
The latest version is v16, so I'd also like to create a follow-up PR to update again.

All definitions stayed the same for v15, only the structure of the package changed: All its exports have been moved to separate files, so I created corresponding `.d.ts `files and updated the tests.
I defined common interfaces in `index.d.ts` and for convenience they are exported again in the files that use them.

This is my first PR, so I hope I followed the process correctly.
Let me know if something needs to be changed 🙂

sheetalkamat commented on types/graphql-upload/index.d.ts:
```diff
@@ -1,53 +1,27 @@
+// TypeScript Version: 4.8
 
-/* tslint:disable:no-unnecessary-generics */
+import { ReadStream } from 'fs-capacitor';
```
This needs to be empty except for metadata since the library doesnt have index.js so having types in index.d.ts doesnt represent correct structure of package.

rdsfj replied:
I did not manage to leave this file completely empty because of the `no-useless-files` linting rule.
If I disable the rule at the very top of the file, the meta data parser breaks because it expects `// Type definitions` in the first line. Disabling the rule later in the file will not satisfy tslint either, it will report the error anyway.
I found no other package with an empty `index.d.ts` and I also didn't want to disable the rule in `tslint.json`.
To solve this, I added a single definition to make tslint happy. Not super nice, so let me know if there is a better way to solve this.

sheetalkamat commented on types/graphql-upload/Upload.d.ts:
```diff
@@ -0,0 +1,8 @@
+import { FileUpload } from './';
+
+export { FileUpload };
```
This should not be reexported as the upload.mjs doesnt have any other reexports

rdsfj replied:
I have now duplicated the types that were previously in `index.d.ts` into each of the files that need them.
I'm not really happy with this, but I see no way to have common types in a separate file and at the same time following the structure of the original package.
Each file exports the types it defines as they are useful when consuming the library, for example to annotate variable and function types.

sheetalkamat commented on types/graphql-upload/graphqlUploadKoa.d.ts:
```diff
@@ -0,0 +1,8 @@
+import { DefaultContext, DefaultState, Middleware } from 'koa';
+import { UploadOptions } from './';
+
+export { UploadOptions };
```
this should not be reexported.

sheetalkamat commented on types/graphql-upload/processRequest.d.ts:
```diff
@@ -0,0 +1,10 @@
+import { IncomingMessage, ServerResponse } from 'http';
+import { GraphQLOperation, UploadOptions } from './';
+
+export { GraphQLOperation, UploadOptions };
```
No reexport here either

andrewbranch commented on types/graphql-upload/index.d.ts:
```diff
@@ -1,53 +1,10 @@
-}
+// Linter does not accept files without definitions. Disabling the rule at
+// the top of the file doesn't work either: It breaks the meta data parser.
+type Dummy = never;
```
Because there are no imports or exports in this file, this type is a global, visible to everyone in their completion lists if they have these types installed 😅. Does `export {}` not satisfy the linter?

andrewbranch commented:
@rdsfj it looks like there’s one more thing that’s needed here. The implementation package at v15.0.2 has an export map defined in its package.json:

```json
"exports": {
  "./GraphQLUpload.js": "./GraphQLUpload.js",
  "./graphqlUploadExpress.js": "./graphqlUploadExpress.js",
  "./graphqlUploadKoa.js": "./graphqlUploadKoa.js",
  "./package.json": "./package.json",
  "./processRequest.js": "./processRequest.js",
  "./Upload.js": "./Upload.js"
 },
```

You will need to make a package.json that mirrors this structure:

```jsonc
{
  "private": true,
  "exports": {
    "./GraphQLUpload.js": {
      "types": "./GraphQLUpload.d.ts"
    },
    // etc...
}
```

The presence of this export map actually prevents any subpaths not explicitly listed in the keys from resolving. Without it, a CommonJS module in Node or a module in a bundler would be able to require/import from `"graphql-upload/GraphQLUpload"`, leaving off the `.js`. But because the package defines the export map, this module specifier will not resolve at runtime/bundle-time. Accordingly, this typings package should reflect the same restrictions so that TypeScript will know about it and give the appropriate resolution errors at compile time (as well as provide correct auto-imports).

Thanks!

rdsfj commented:
@andrewbranch Sure thing, I added the `exports` as you suggested. I wasn't sure about `package.json`, so I added `"./package.json": "./package.json"` because I saw that some other packages in this repo do that as well.

rdsfj commented:
Ready to merge

samkit5495 commented:
After this change, I am not able to import GraphQLUpload type
Imported in this way:
```
import GraphQLUpload from 'graphql-upload/GraphQLUpload';
```
Error:
```
Error [ERR_PACKAGE_PATH_NOT_EXPORTED]: Package subpath './GraphQLUpload' is not defined by "exports" in /Users/samkit5495/Documents/Projects/WooChat.io/woochat-backend/node_modules/graphql-upload/package.json
    at new NodeError (internal/errors.js:322:7)
    at throwExportsNotFound (internal/modules/esm/resolve.js:332:9)
    at packageExportsResolve (internal/modules/esm/resolve.js:565:3)
    at resolveExports (internal/modules/cjs/loader.js:478:36)
    at Function.Module._findPath (internal/modules/cjs/loader.js:518:31)
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:917:27)
    at Function.Module._load (internal/modules/cjs/loader.js:774:27)
    at Module.require (internal/modules/cjs/loader.js:1003:19)
    at require (internal/modules/cjs/helpers.js:107:18)
```
